src/vs/workbench/contrib/debug/common/abstractDebugAdapter.ts

```diff
@@ -110,29 +111,43 @@ export abstract class AbstractDebugAdapter implements IDebugAdapter {
 			this.messageCallback(message);
 		}
 		else {
-			switch (message.type) {
-				case 'event':
-					if (this.eventCallback) {
-						this.eventCallback(<DebugProtocol.Event>message);
-					}
-					break;
-				case 'request':
-					if (this.requestCallback) {
-						this.requestCallback(<DebugProtocol.Request>message);
-					}
-					break;
-				case 'response':
-					const response = <DebugProtocol.Response>message;
-					const clb = this.pendingRequests.get(response.request_seq);
-					if (clb) {
-						this.pendingRequests.delete(response.request_seq);
-						clb(response);
-					}
-					break;
+			// Artificially queueing protocol messages guarantees that any microtasks for
+			// previous message finish before next message is processed. This is essential
+			// to guarantee ordering when using promises anywhere along the call path.
+			this.queue.push(message);
+			if (this.queue.length === 1) {
+				setTimeout(() => this.processQueue(), 0);
 			}
 		}
 	}
 
+	private processQueue(): void {
+		const message = this.queue!.shift()!;
```

Comment:
I do not understand why you need `this.queue!`
`this.queue` is never null and you properly initiazlie it to an empty array. So you should be always safe to access `this.queue`.

As for the ! after the shift, I would prefer to just check if the message is there by `if !(message) {return}`. Just looks safer to me.